runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,351 @@
+local api = vim.api
+local lsp = vim.lsp
+local util = lsp.util
+local ms = lsp.protocol.Methods
+
+local document_color_ns = api.nvim_create_namespace('nvim.lsp.document_color')
+local document_color_augroup = api.nvim_create_augroup('nvim.lsp.document_color', {})
+
+local M = {}
+
+--- @class (private) vim.lsp.document_color.HighlightInfo
+--- @field hex_code string Resolved HEX color
+--- @field range Range4 Range of the highlight
+--- @field hl_group? string Highlight group name. Won't be present if the style is a custom function.
+
+--- @class (private) vim.lsp.document_color.BufState
+--- @field enabled boolean Whether document_color is enabled for the current buffer
+--- @field buf_version? integer Buffer version for which the color ranges correspond to
+--- @field applied_version? table<integer, integer> (client_id -> buffer version) Last buffer version for which we applied color ranges
+--- @field hl_info? table<integer, vim.lsp.document_color.HighlightInfo[]> (client_id -> color highlights) Processed highlight information
+
+--- @type table<integer, vim.lsp.document_color.BufState>
```

Comment:
```suggestion
--- @field applied_version table<integer, integer> (client_id -> buffer version) Last buffer version for which we applied color ranges
--- @field hl_info table<integer, vim.lsp.document_color.HighlightInfo[]?> (client_id -> color highlights) Processed highlight information

--- @type table<integer, vim.lsp.document_color.BufState?>
```

Both `applied_version` and `hl_info` are nested tables; we could make them required to ensure they exist if `bufstate` exists, to avoid another nil-check.

We could modify `table<K, V>` to `table<K, V?`> for a stronger type annotation.